File: ical_test.go
package main

import (
	"strings"
	"testing"
	"time"
)

// icalNow is a fixed instant used to make DTSTAMP deterministic. It is given in a
// non-UTC zone to confirm buildICS converts the stamp to UTC.
var icalNow = time.Date(2026, 6, 12, 12, 0, 0, 0, time.FixedZone("UTC+2", 2*60*60))

// lines splits an ICS document into its logical CRLF-separated lines and asserts
// CRLF endings were used throughout.
func lines(t *testing.T, ics string) []string {
	t.Helper()
	if strings.Contains(strings.ReplaceAll(ics, "\r\n", ""), "\n") {
		t.Fatalf("found a bare LF not part of a CRLF pair:\n%q", ics)
	}
	return strings.Split(strings.TrimSuffix(ics, "\r\n"), "\r\n")
}

func countLine(ls []string, want string) int {
	n := 0
	for _, l := range ls {
		if l == want {
			n++
		}
	}
	return n
}

func hasLine(ls []string, want string) bool {
	return countLine(ls, want) > 0
}

func TestBuildICSStructureAndEvents(t *testing.T) {
	tasks := []*Task{
		{
			Name:        "Linear Algebra",
			Description: "Eigenvalues",
			Sessions: []*Session{
				{ID: "s1", Date: "2026-06-14"},
				{ID: "s2", Date: "2026-06-20", Done: true}, // done: excluded
			},
		},
		{
			Name:     "Archived",
			Archived: true,
			Sessions: []*Session{{ID: "s3", Date: "2026-06-15"}}, // archived: excluded
		},
		{
			Name:     "History",
			Sessions: []*Session{{ID: "s4", Date: "2026-06-13"}}, // no description
		},
	}

	ics := buildICS(tasks, icalNow)
	ls := lines(t, ics)

	// Calendar wrapper.
	if ls[0] != "BEGIN:VCALENDAR" || ls[len(ls)-1] != "END:VCALENDAR" {
		t.Fatalf("missing VCALENDAR wrapper:\n%s", ics)
	}
	for _, want := range []string{"VERSION:2.0", "PRODID:" + iCalProdID, "CALSCALE:GREGORIAN", "X-WR-CALNAME:Study Planner"} {
		if !hasLine(ls, want) {
			t.Errorf("missing header line %q", want)
		}
	}

	// Exactly two events: the pending sessions s1 and s4 (done + archived dropped).
	if got := countLine(ls, "BEGIN:VEVENT"); got != 2 {
		t.Fatalf("got %d VEVENTs, want 2:\n%s", got, ics)
	}
	for _, id := range []string{"s1", "s4"} {
		if !hasLine(ls, "UID:"+id+"@study-planner") {
			t.Errorf("missing UID for %q", id)
		}
	}
	for _, id := range []string{"s2", "s3"} {
		if hasLine(ls, "UID:"+id+"@study-planner") {
			t.Errorf("UID for excluded session %q should not appear", id)
		}
	}

	// DTSTAMP is the supplied instant converted to UTC (12:00 in UTC+2 → 10:00Z).
	if !hasLine(ls, "DTSTAMP:20260612T100000Z") {
		t.Errorf("DTSTAMP not converted to UTC:\n%s", ics)
	}

	// All-day event: DTEND is the day after DTSTART.
	if !hasLine(ls, "DTSTART;VALUE=DATE:20260614") || !hasLine(ls, "DTEND;VALUE=DATE:20260615") {
		t.Errorf("all-day DTSTART/DTEND wrong:\n%s", ics)
	}
	if !hasLine(ls, "SUMMARY:Study: Linear Algebra") {
		t.Errorf("missing summary:\n%s", ics)
	}
	if !hasLine(ls, "DESCRIPTION:Eigenvalues") {
		t.Errorf("missing description:\n%s", ics)
	}
}

func TestBuildICSSortedByDate(t *testing.T) {
	tasks := []*Task{{
		Name: "T",
		Sessions: []*Session{
			{ID: "late", Date: "2026-07-01"},
			{ID: "early", Date: "2026-06-01"},
		},
	}}
	ls := lines(t, buildICS(tasks, icalNow))
	var order []string
	for _, l := range ls {
		if strings.HasPrefix(l, "UID:") {
			order = append(order, l)
		}
	}
	if len(order) != 2 || order[0] != "UID:early@study-planner" {
		t.Fatalf("events not date-sorted: %v", order)
	}
}

func TestBuildICSEscaping(t *testing.T) {
	tasks := []*Task{{
		Name:        "Maths, Physics; Chemistry\\Bio",
		Description: "line one\nline two",
		Sessions:    []*Session{{ID: "s1", Date: "2026-06-14"}},
	}}
	ics := buildICS(tasks, icalNow)
	if !strings.Contains(ics, `SUMMARY:Study: Maths\, Physics\; Chemistry\\Bio`) {
		t.Errorf("summary special chars not escaped:\n%s", ics)
	}
	if !strings.Contains(ics, `DESCRIPTION:line one\nline two`) {
		t.Errorf("newline in description not escaped:\n%s", ics)
	}
}

func TestBuildICSFoldsLongLines(t *testing.T) {
	long := strings.Repeat("A", 200)
	tasks := []*Task{{
		Name:     long,
		Sessions: []*Session{{ID: "s1", Date: "2026-06-14"}},
	}}
	for _, l := range lines(t, buildICS(tasks, icalNow)) {
		// Continuation lines begin with a space; every physical line must stay
		// within the 75-octet limit.
		if len(l) > 75 {
			t.Fatalf("line exceeds 75 octets (%d): %q", len(l), l)
		}
	}
}

func TestBuildICSSkipsMalformedDate(t *testing.T) {
	tasks := []*Task{{
		Name: "T",
		Sessions: []*Session{
			{ID: "bad", Date: "not-a-date"},
			{ID: "ok", Date: "2026-06-14"},
		},
	}}
	ls := lines(t, buildICS(tasks, icalNow))
	if got := countLine(ls, "BEGIN:VEVENT"); got != 1 {
		t.Fatalf("got %d events, want 1 (malformed date skipped)", got)
	}
	if !hasLine(ls, "UID:ok@study-planner") {
		t.Error("valid session missing")
	}
}

func TestBuildICSEmpty(t *testing.T) {
	ls := lines(t, buildICS(nil, icalNow))
	if got := countLine(ls, "BEGIN:VEVENT"); got != 0 {
		t.Fatalf("expected no events, got %d", got)
	}
	if ls[0] != "BEGIN:VCALENDAR" || ls[len(ls)-1] != "END:VCALENDAR" {
		t.Error("empty calendar should still have wrapper")
	}
}

File: frontend/src/lib/backend/local/ical.test.ts
// Ports of ical_test.go — the RFC 5545 output spec, including the UTC DTSTAMP
// conversion and the 75-octet folding limit.
import { describe, expect, it } from 'vitest';
import { buildICS, ICAL_PROD_ID } from './ical';
import type { Session, Task } from './models';

// Go's icalNow is 12:00 in UTC+2, i.e. 10:00Z — buildICS must stamp in UTC.
const ICAL_NOW = new Date(Date.UTC(2026, 5, 12, 10, 0, 0));

function mkSession(id: string, date: string, done = false): Session {
  return { id, date, done };
}

function mkTask(partial: Partial<Task> & { name: string }): Task {
  return {
    id: partial.name,
    description: '',
    color: '',
    subjectId: '',
    tags: [],
    archived: false,
    adaptive: false,
    order: 0,
    createdAt: new Date(0).toISOString(),
    sessions: [],
    ...partial,
  };
}

// lines splits an ICS document into its logical CRLF-separated lines, asserting
// CRLF endings were used throughout.
function lines(ics: string): string[] {
  expect(ics.replace(/\r\n/g, ''), 'found a bare LF not part of a CRLF pair').not.toContain('\n');
  return ics.replace(/\r\n$/, '').split('\r\n');
}

const countLine = (ls: string[], want: string) => ls.filter((l) => l === want).length;
const hasLine = (ls: string[], want: string) => countLine(ls, want) > 0;

describe('buildICS', () => {
  it('emits one all-day VEVENT per pending session of active tasks', () => {
    const tasks = [
      mkTask({
        name: 'Linear Algebra',
        description: 'Eigenvalues',
        sessions: [
          mkSession('s1', '2026-06-14'),
          mkSession('s2', '2026-06-20', true), // done: excluded
        ],
      }),
      mkTask({
        name: 'Archived',
        archived: true,
        sessions: [mkSession('s3', '2026-06-15')], // archived: excluded
      }),
      mkTask({
        name: 'History',
        sessions: [mkSession('s4', '2026-06-13')], // no description
      }),
    ];

    const ics = buildICS(tasks, ICAL_NOW);
    const ls = lines(ics);

    // Calendar wrapper.
    expect(ls[0]).toBe('BEGIN:VCALENDAR');
    expect(ls[ls.length - 1]).toBe('END:VCALENDAR');
    for (const want of [
      'VERSION:2.0',
      'PRODID:' + ICAL_PROD_ID,
      'CALSCALE:GREGORIAN',
      'X-WR-CALNAME:Study Planner',
    ]) {
      expect(hasLine(ls, want), want).toBe(true);
    }

    // Exactly two events: the pending sessions s1 and s4.
    expect(countLine(ls, 'BEGIN:VEVENT')).toBe(2);
    for (const id of ['s1', 's4']) expect(hasLine(ls, `UID:${id}@study-planner`), id).toBe(true);
    for (const id of ['s2', 's3']) expect(hasLine(ls, `UID:${id}@study-planner`), id).toBe(false);

    // DTSTAMP is the supplied instant in UTC.
    expect(hasLine(ls, 'DTSTAMP:20260612T100000Z')).toBe(true);

    // All-day event: DTEND is the day after DTSTART.
    expect(hasLine(ls, 'DTSTART;VALUE=DATE:20260614')).toBe(true);
    expect(hasLine(ls, 'DTEND;VALUE=DATE:20260615')).toBe(true);
    expect(hasLine(ls, 'SUMMARY:Study: Linear Algebra')).toBe(true);
    expect(hasLine(ls, 'DESCRIPTION:Eigenvalues')).toBe(true);
  });

  it('sorts events by date', () => {
    const tasks = [
      mkTask({
        name: 'T',
        sessions: [mkSession('late', '2026-07-01'), mkSession('early', '2026-06-01')],
      }),
    ];
    const order = lines(buildICS(tasks, ICAL_NOW)).filter((l) => l.startsWith('UID:'));
    expect(order).toHaveLength(2);
    expect(order[0]).toBe('UID:early@study-planner');
  });

  it('escapes TEXT specials and newlines', () => {
    const tasks = [
      mkTask({
        name: 'Maths, Physics; Chemistry\\Bio',
        description: 'line one\nline two',
        sessions: [mkSession('s1', '2026-06-14')],
      }),
    ];
    const ics = buildICS(tasks, ICAL_NOW);
    expect(ics).toContain('SUMMARY:Study: Maths\\, Physics\\; Chemistry\\\\Bio');
    expect(ics).toContain('DESCRIPTION:line one\\nline two');
  });

  it('folds long lines to 75 octets', () => {
    const encoder = new TextEncoder();
    const tasks = [mkTask({ name: 'A'.repeat(200), sessions: [mkSession('s1', '2026-06-14')] })];
    for (const l of lines(buildICS(tasks, ICAL_NOW))) {
      expect(encoder.encode(l).length, l).toBeLessThanOrEqual(75);
    }
  });

  it('skips sessions with malformed dates', () => {
    const tasks = [
      mkTask({
        name: 'T',
        sessions: [mkSession('bad', 'not-a-date'), mkSession('ok', '2026-06-14')],
      }),
    ];
    const ls = lines(buildICS(tasks, ICAL_NOW));
    expect(countLine(ls, 'BEGIN:VEVENT')).toBe(1);
    expect(hasLine(ls, 'UID:ok@study-planner')).toBe(true);
  });

  it('renders an empty calendar with the wrapper intact', () => {
    const ls = lines(buildICS([], ICAL_NOW));
    expect(countLine(ls, 'BEGIN:VEVENT')).toBe(0);
    expect(ls[0]).toBe('BEGIN:VCALENDAR');
    expect(ls[ls.length - 1]).toBe('END:VCALENDAR');
  });
});
